Semantic Convention generation tooling

Aneurysm9 opened the pull request:
This adds a tool to use the `otel/semconvgen` container image to generate resource and trace semantic conventions from the specification.  It is not yet integrated with build processes and must be manually run, but it can produce constant and variable declarations with reasonable godoc.

Fixes #1879.

MadVikingGod commented:
Just because I couldn't tell I created a gist of the differences between keys: https://gist.github.com/MadVikingGod/9d108870779419171e7edecafe86e323

This isn't a comment on if this is appropriate or not, just so others don't have to try and work this out.

It looks like a few new Keys:
- AWS*
- DBCassandra
- FaaS*
- MessagingKafka*
- other misc keys

Major change of
- `K8s*` to `K8S*`
- `CodeLineNumberKey` to `CodeLinenoKey`
- `DBRedisDBIndexKey` to `DBRedisDatabaseIndexKey`

And the removal of 
- RPCMessage*
- MessagingRabbit
- `RPCNameKey`

Aneurysm9 commented:
Thanks for doing that analysis @MadVikingGod!  I've added replacements to fix the `K8S*` (though I wonder whether that's right), `Redis`, and `LineNumber` key names.

There were some `FaaS*` keys previously, though I think they were primarily in `resource.go` and new ones were added in `trace.go`.  All of the added and removed keys are according to the conventions as they currently exist in the specification repo.  If we want to restore the removed keys we'll need to get them into the spec and re-generate these files.

pellared commented on internal/tools/semconv-gen/generator.go:
```diff
@@ -0,0 +1,300 @@
+	"Lineno":        "LineNumber",
+}
+
+func fixInitialisms(fn string) error {
```
there is quite a logic here 😉 
what do you think of adding some tests here using some input and expected "golden files"  kept in `testdata` directory?
It would help fix potential problems with the replacement mechanism in a more automated and faster way.

Aneurysm9 replied:
The fixup logic is pretty simple, I think.  I've added some comments and separated the `go fmt` invocation.  I view the spec as the "golden files" and the current output as the target for testing any refactoring.  It should always be possible to remove the generated output, regenerate them, and check that git doesn't think anything has changed if testing a refactoring.  If testing a change intended to induce a change in the output, the same thing applies.  Git's diff mechanism can let us know if we've done what we expected to do, though we shouldn't be changing generated outputs once they're shipped so I don't think it will be an issue in practice.

pellared commented on internal/tools/semconv-gen/generator.go:
```diff
@@ -0,0 +1,300 @@
+	"regexp"
+	"strings"
+
+	flag "github.com/spf13/pflag"
```
Do we really need to use https://github.com/spf13/pflag ? 
I feel that it is feature creeping and we would need to update this dependency. 
I would prefer to use just `flag`. 
Also, it is safer to just use the standard library from a security standpoint.

Aneurysm9 replied:
It's an internal tool that doesn't ship in any end-user code.  I prefer the POSIX-style flags.  The pflag package is stable and widely used, I see no issue with using it here.

pellared commented on internal/tools/semconv-gen/generator.go:
```diff
@@ -0,0 +1,300 @@
+	flag "github.com/spf13/pflag"
+)
+
+var outputFn = flag.StringP("filename", "f", "", "Filename for template output. If not specified 'basename(inputPath).go' will be used.")
+var inputPath = flag.StringP("input", "i", "", "Path to semantic convention definition YAML")
+var outputPath = flag.StringP("output", "o", "semconv", "Path to output target. Must be either an absolute path or relative to the repository root.")
+var templateFn = flag.StringP("template", "t", "template.j2", "Template filename")
+var containerImage = flag.StringP("container", "c", "otel/semconvgen", "Container image ID")
+
+func main() {
+	flag.Parse()
+
+	cfg, err := buildConfig()
```
Can we get rid of globals and just to something more or less like:

```go
func main() {
   cfg := &config{}
   flag.StringVarP(&cfg.outputPath, "filename", "f", "", "Filename for template output. If not specified 'basename(inputPath).go' will be used."
   ...
   validCfg, err := buildConfig(cfg)
}
```

pellared commented on CHANGELOG.md:
```diff
@@ -24,6 +24,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - The `Status` type was added to the `go.opentelemetry.io/otel/sdk/trace` package to represent the status of a span. (#1874)
 - The `SpanStub` type and its associated functions were added to the `go.opentelemetry.io/otel/sdk/trace/tracetest` package.
   This type can be used as a testing replacement for the `SpanSnapshot` that was removed from the `go.opentelemetry.io/otel/sdk/trace` package. (#1873)
+- Semantic convention generation tooling has been added. (#1891)
```
Is this something that OTel API/SDK users should be aware of? If so then probably some docs for `semconv-gen` should be added.

Aneurysm9 replied:
No, not really.  It's an internal too meant for maintainer use.  This line can probably be removed.

pellared replied:
Let's remove it then  🚀

pellared reviewed (commented):
Looks nice 👍 Added a just few comments

MadVikingGod reviewed (commented):
Other than finding the repo root magic, I think this is GTG.

MrAlias commented on internal/tools/semconv-gen/generator.go:
```diff
@@ -0,0 +1,300 @@
+		return fmt.Errorf("unable to create output directory: %w", err)
+	}
+
+	err = exec.Command("cp", "-a", cfg.input, path.Join(tmpDir, "input")).Run()
```
nit

```suggestion
	err = exec.Command("cp", "-a", cfg.input, inputPath).Run()
```

MrAlias commented on semconv/resource.go:
```diff
@@ -12,246 +12,734 @@
+	// Azure Virtual Machines
+	CloudPlatformAzureVM = CloudPlatformKey.String("azure_vm")
+	// Azure Container Instances
+	CloudPlatformAzureContainerinstances = CloudPlatformKey.String("azure_container_instances")
```
Should this be `CloudPlatformAzureContainerInstances`?

MrAlias commented on semconv/resource.go:
```diff
@@ -12,246 +12,734 @@
+	// Azure Functions
+	CloudPlatformAzureFunctions = CloudPlatformKey.String("azure_functions")
+	// Azure App Service
+	CloudPlatformAzureAppservice = CloudPlatformKey.String("azure_app_service")
```
```suggestion
	CloudPlatformAzureAppService = CloudPlatformKey.String("azure_app_service")
```

MrAlias commented on semconv/resource.go:
```diff
@@ -12,246 +12,734 @@
+	// Azure App Service
+	CloudPlatformAzureAppservice = CloudPlatformKey.String("azure_app_service")
+	// Google Cloud Compute Engine (GCE)
+	CloudPlatformGCPComputeengine = CloudPlatformKey.String("gcp_compute_engine")
```
```suggestion
	CloudPlatformGCPComputeEngine = CloudPlatformKey.String("gcp_compute_engine")
```

MrAlias commented on semconv/resource.go:
```diff
@@ -12,246 +12,734 @@
+	// Google Cloud Compute Engine (GCE)
+	CloudPlatformGCPComputeengine = CloudPlatformKey.String("gcp_compute_engine")
+	// Google Cloud Run
+	CloudPlatformGCPCloudrun = CloudPlatformKey.String("gcp_cloud_run")
```
```suggestion
	CloudPlatformGCPCloudRun = CloudPlatformKey.String("gcp_cloud_run")
```

MrAlias commented on semconv/resource.go:
```diff
@@ -12,246 +12,734 @@
+	// Google Cloud Run
+	CloudPlatformGCPCloudrun = CloudPlatformKey.String("gcp_cloud_run")
+	// Google Cloud Kubernetes Engine (GKE)
+	CloudPlatformGCPKubernetesengine = CloudPlatformKey.String("gcp_kubernetes_engine")
```
```suggestion
	CloudPlatformGCPKubernetesEngine = CloudPlatformKey.String("gcp_kubernetes_engine")
```

MrAlias commented on semconv/resource.go:
```diff
@@ -12,246 +12,734 @@
+	// Google Cloud Kubernetes Engine (GKE)
+	CloudPlatformGCPKubernetesengine = CloudPlatformKey.String("gcp_kubernetes_engine")
+	// Google Cloud Functions (GCF)
+	CloudPlatformGCPCloudfunctions = CloudPlatformKey.String("gcp_cloud_functions")
```
```suggestion
	CloudPlatformGCPCloudFunctions = CloudPlatformKey.String("gcp_cloud_functions")
```

MrAlias commented on semconv/resource.go:
```diff
@@ -12,246 +12,734 @@
+	// Google Cloud Functions (GCF)
+	CloudPlatformGCPCloudfunctions = CloudPlatformKey.String("gcp_cloud_functions")
+	// Google Cloud App Engine (GAE)
+	CloudPlatformGCPAppengine = CloudPlatformKey.String("gcp_app_engine")
```
```suggestion
	CloudPlatformGCPAppEngine = CloudPlatformKey.String("gcp_app_engine")
```

MrAlias commented on semconv/resource.go:
```diff
@@ -12,246 +12,734 @@
+	// Note: See the [log group ARN format
+	// documentation](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/iam-
+	// access-control-overview-cwl.html#CWL_ARN_Format).
+	AWSLogGroupArnsKey = attribute.Key("aws.log.group.arns")
```
Isn't ARN an acronym? Maybe it needs to be added to the list.

MrAlias commented on semconv/resource.go:
```diff
@@ -12,246 +12,734 @@
+	// Type: string
+	// Required: No
+	// Examples: '275ecb36-5aa8-4c2a-9c47-d8bb681b9aff'
+	K8SCronjobUIDKey = attribute.Key("k8s.cronjob.uid")
```
Can we add an override here to capitalize the two worded "cron job"?

```suggestion
	K8SCronJobUIDKey = attribute.Key("k8s.cronjob.uid")
```

MrAlias commented on semconv/trace.go:
```diff
@@ -12,365 +12,1016 @@
+	// SAP MaxDB
+	DBSystemMaxDB = DBSystemKey.String("maxdb")
+	// SAP HANA
+	DBSystemHanadb = DBSystemKey.String("hanadb")
```
```suggestion
	DBSystemHanaDB = DBSystemKey.String("hanadb")
```

right?

MrAlias commented on semconv/trace.go:
```diff
@@ -12,365 +12,1016 @@
+	// Type: string
+	// Required: No
+	// Examples: 'us-west-2'
+	DBCassandraCoordinatorDcKey = attribute.Key("db.cassandra.coordinator.dc")
```
```suggestion
	DBCassandraCoordinatorDCKey = attribute.Key("db.cassandra.coordinator.dc")
```

MrAlias commented on semconv/trace.go:
```diff
@@ -12,365 +12,1016 @@
+	// Type: Enum
+	// Required: Always
+	// Examples: 0, 1, 16
+	RPCGrpcStatusCodeKey = attribute.Key("rpc.grpc.status_code")
```
```suggestion
	RPCGRPCStatusCodeKey = attribute.Key("rpc.grpc.status_code")
```

Aneurysm9 commented on semconv/http.go:
```diff
@@ -25,6 +25,11 @@ import (
 	"go.opentelemetry.io/otel/codes"
 )
 
+var (
+	HTTPSchemeHTTP  = HTTPSchemeKey.String("http")
+	HTTPSchemeHTTPS = HTTPSchemeKey.String("https")
+)
+
```
This was just moved out of the other files.  It doesn't exist in the conventions and I'm not actually sure we should be introducing it, but it was used in some of the helpers here and I figured it was easier to move closer to its use and we can discuss whether to actually keep them separately if needed.

Aneurysm9 commented on semconv/resource.go:
```diff
@@ -12,246 +12,734 @@
-	HostImageVersionKey = attribute.Key("host.image.version")
+var (
+	// Microsoft Windows
+	OSTypeWindows = OSTypeKey.String("WINDOWS")
+	// Linux
+	OSTypeLinux = OSTypeKey.String("LINUX")
+	// Apple Darwin
+	OSTypeDarwin = OSTypeKey.String("DARWIN")
+	// FreeBSD
+	OSTypeFreeBSD = OSTypeKey.String("FREEBSD")
+	// NetBSD
+	OSTypeNetBSD = OSTypeKey.String("NETBSD")
+	// OpenBSD
+	OSTypeOpenBSD = OSTypeKey.String("OPENBSD")
+	// DragonFly BSD
+	OSTypeDragonflyBSD = OSTypeKey.String("DRAGONFLYBSD")
+	// HP-UX (Hewlett Packard Unix)
+	OSTypeHPUX = OSTypeKey.String("HPUX")
+	// AIX (Advanced Interactive eXecutive)
+	OSTypeAIX = OSTypeKey.String("AIX")
+	// Oracle Solaris
+	OSTypeSolaris = OSTypeKey.String("SOLARIS")
+	// IBM z/OS
+	OSTypeZOS = OSTypeKey.String("ZOS")
```
Good catch indeed, my spec repo was behind the times.  I'll update this PR generating from the `v1.3.0` spec tag.

Aneurysm9 commented on semconv/trace.go:
```diff
@@ -12,365 +12,1016 @@
-	NetTransportUnix   = NetTransportKey.String("Unix")
-	NetTransportPipe   = NetTransportKey.String("pipe")
+	// IP.TCP
+	NetTransportTCP = NetTransportKey.String("IP.TCP")
+	// IP.UDP
+	NetTransportUDP = NetTransportKey.String("IP.UDP")
+	// Another IP-based protocol
+	NetTransportIP = NetTransportKey.String("IP")
+	// Unix Domain socket. See below
+	NetTransportUnix = NetTransportKey.String("Unix")
+	// Named or anonymous pipe. See note below
+	NetTransportPipe = NetTransportKey.String("pipe")
+	// In-process communication
 	NetTransportInProc = NetTransportKey.String("inproc")
-	NetTransportOther  = NetTransportKey.String("other")
+	// Something else (non IP-based)
+	NetTransportOther = NetTransportKey.String("other")
 )
```
I think that's just me being bad and generating from outdated spec.  Will be updated with the next push.

MadVikingGod commented on internal/tools/semconv-gen/generator.go:
```diff
@@ -0,0 +1,300 @@
+	return nil
+}
+
+func findRepoRoot() (string, error) {
+	start, err := os.Getwd()
+	if err != nil {
+		return "", err
+	}
+
+	dir := start
+	for {
+		_, err := os.Stat(filepath.Join(dir, ".git"))
+		if errors.Is(err, os.ErrNotExist) {
+			dir = filepath.Dir(dir)
+			// From https://golang.org/pkg/path/filepath/#Dir:
+			// The returned path does not end in a separator unless it is the root directory.
+			if strings.HasSuffix(dir, string(filepath.Separator)) {
+				return "", fmt.Errorf("unable to find git repository enclosing working dir %s", start)
+			}
+			continue
+		}
+
+		if err != nil {
+			return "", err
+		}
+
+		return dir, nil
+	}
+}
```
That is understandable, and while I don't like it, I think it makes sense.

For tests, I don't think this addresses the automation of generating these at release.  That is where I would expect the "test" to live.  And it would be much like the mod check from `check-clean-work-tree` in our CI.

To resolve this could we either add to our documentation (if we have it) what to do with this tool as part of a release, or start a ticket to create that document and include it as one of the action items in that?

Aneurysm9 replied:
I've added some notes to the `RELEASING.md` to provide guidance on when and how to use it.  I think as we  work through #1890 and how we want to handle versioning of the generated semantic conventions the docs will need to be updated, but it should be a good start.

Aneurysm9 replied:
I've created #1907 to track figuring out how to automate more of this process.

pellared commented on internal/tools/semconv-gen/generator.go:
```diff
@@ -0,0 +1,319 @@
+	if err != nil {
+		fmt.Println(err)
+		flag.Usage()
+		os.Exit(-1)
```
nit: this is not POSIX compliant. TBH I think that have never seen negative exit codes 😉 
 https://tldp.org/LDP/abs/html/exitcodes.html

AFAIK `2` is generally used for bad input

```suggestion
		os.Exit(2)
```

pellared replied:
This is what `flag` package does: https://github.com/golang/go/blob/485474d204297ce9ff2b7f6c2a2b0e048c61fb8e/src/flag/flag.go#L317

Aneurysm9 replied:
Shells have a slightly different range for valid exit codes, reserving 128-255.  The docs you link also specify that `2` is intended to represent "[m]isuse of shell builtins (according to Bash documentation)".  This isn't a shell utility though, so the relevant portion of the POSIX spec would be https://pubs.opengroup.org/onlinepubs/009695399/functions/exit.html.  This specifies that only the lower 8 bits of the exit code will be available, leading to an [exit code of 255](https://play.golang.org/p/xH8dT4uNg-z).  Maybe I'm old and times have changed out from under me, but `exit -1` is how I've always signaled abnormal termination.

pellared replied:
Resolve :-)
